cpp/dolfin/fem/ElementDofLayout.cpp

```diff
@@ -5,46 +5,82 @@
 // SPDX-License-Identifier:    LGPL-3.0-or-later
 
 #include "ElementDofLayout.h"
+#include "ReferenceCellTopology.h"
+#include <array>
 #include <dolfin/mesh/CellType.h>
-#include <ufc.h>
+#include <map>
+#include <set>
 
 using namespace dolfin;
 using namespace dolfin::fem;
 
 //-----------------------------------------------------------------------------
 ElementDofLayout::ElementDofLayout(
-    int block_size, std::vector<std::vector<std::vector<int>>> entity_dofs,
-    std::vector<std::vector<std::vector<int>>> entity_closure_dofs,
+    int block_size, std::vector<std::vector<std::set<int>>> entity_dofs,
     std::vector<int> parent_map,
-    std::vector<std::shared_ptr<ElementDofLayout>> sub_dofmaps)
+    std::vector<std::shared_ptr<ElementDofLayout>> sub_dofmaps,
+    const mesh::CellType& cell_type)
     : _parent_map(parent_map), _block_size(block_size), _num_dofs(0),
-      _entity_dofs(entity_dofs), _entity_closure_dofs(entity_closure_dofs),
-      _sub_dofmaps(sub_dofmaps)
+      _entity_dofs(entity_dofs), _sub_dofmaps(sub_dofmaps)
 {
   // TODO: Handle global support dofs
 
-  // TODO: Add reference cell class to allow entity_closure_dofs to be
-  //       removed as argument to constructor
+  dolfin::CellType _cell = dolfin::CellType::point;
+  if (cell_type.cell_type() == mesh::CellType::Type::interval)
+    _cell = dolfin::CellType::interval;
+  else if (cell_type.cell_type() == mesh::CellType::Type::triangle)
+    _cell = dolfin::CellType::triangle;
+  else if (cell_type.cell_type() == mesh::CellType::Type::quadrilateral)
+    _cell = dolfin::CellType::quadrilateral;
+  else if (cell_type.cell_type() == mesh::CellType::Type::tetrahedron)
+    _cell = dolfin::CellType::tetrahedron;
+  else if (cell_type.cell_type() == mesh::CellType::Type::hexahedron)
+    _cell = dolfin::CellType::hexahedron;
+  else
+    throw std::runtime_error("Ooops");
+
+  const int* num_entities = ReferenceCellTopology::num_entities(_cell);
+  assert(num_entities);
+
+  // Compute closure entities
+  // [dim, entity] -> closure{sub_dim, (sub_entities)}
+  std::map<std::array<int, 2>, std::map<int, std::set<int>>> entity_closure
```

Comment:
Not sure the second `map` is needed, as sub_dim will be 0, 1, 2 etc. i.e. could just be a vector?

Reply:
Sure - feel free to push a change.